### PR TITLE
mam/messages: trusted channels/endpoints

### LINKED ISSUE #795: mam2: implement mam_add_trusted_channel_id

### LINKED ISSUE #796: mam2: implement mam_add_trusted_endpoint_id

### PR DESCRIPTION
**Pull request by tsvisabo**
Fixes #795 
Fixes #796

### CODE REVIEW

File: mam/mam/message.h [hunk 1]
```diff
@@ -144,13 +145,17 @@ retcode_t mam_msg_write_packet(mam_msg_write_context_t *const ctx,
...
+ * @param trusted_endpoint_pks - The set of trusted endpoint pks [in, out]
  *
  * @return return error code
  */
 retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
                               trits_t *const msg, mam_psk_t_set_t const psks,
                               mam_ntru_sk_t_set_t const ntru_sks,
-                              trits_t msg_id);
+                              trits_t msg_id,
+                              mam_pk_t_set_t *const trusted_channels_pks,
```
Comment (thibault-martinez) [inline mam/mam/message.h:155]: trusted_channel_pks

File: mam/mam/message.h [hunk 2]
```diff
@@ -144,13 +145,17 @@ retcode_t mam_msg_write_packet(mam_msg_write_context_t *const ctx,
  * @param psks - The set of pre-shared-keys[in]
  * @param ntru_sks - The set of the NTRU secret keys[in]
  * @param msg_id - The msg_id[out]
+ * @param trusted_channels_pks - The set of trusted channels pks [in, out]
```
Comment (thibault-martinez) [inline mam/mam/message.h:146]: trusted_channels_pks

File: mam/mam/message.c [hunk 3]
```diff
@@ -712,7 +712,9 @@ retcode_t mam_msg_write_packet(mam_msg_write_context_t *const ctx,
 retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
                               trits_t *const msg, mam_psk_t_set_t const psks,
                               mam_ntru_sk_t_set_t const ntru_sks,
-                              trits_t msg_id) {
+                              trits_t msg_id,
+                              mam_pk_t_set_t *const trusted_channels_pks,
```
Comment (thibault-martinez) [inline mam/mam/message.c:717]: trusted_channels_pks

File: mam/examples/recv.c [hunk 4]
```diff
@@ -37,6 +38,11 @@ int main(int ac, char **av) {
   receive_bundle(av[1], atoi(av[2]), (tryte_t *)av[3], bundle);
 
   mam_psk_t_set_add(&api.psks, &psk);
+  if (ac == 5) {
+    trytes_to_trits((tryte_t *)av[4], chid.pk,
+                    MAM_CHANNEL_ID_SIZE / NUMBER_OF_TRITS_IN_A_TRYTE);
+    mam_pk_t_set_add(&api.trusted_channel_ids, &chid);
```
Comment (thibault-martinez) [inline mam/examples/recv.c]: Can you provide 2 api functions that do these two operations ?
They are actually #796  and #795

File: mam/api/api.c [hunk 5]
```diff
@@ -367,6 +375,9 @@ retcode_t mam_api_create_endpoint(mam_api_t *const api, size_t const height,
     mam_endpoint_destroy(&endpoint);
     return ret;
   }
+  ERR_BIND_RETURN(mam_pk_t_set_add(&api->trusted_channel_ids,
+                                   trits_begin(mam_channel_id(&channel))),
```
Comment (thibault-martinez) [inline mam/api/api.c]: endpoint

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/examples/recv.c (hunk 4 above)
Comment (tsvisabo) [reply mam/examples/recv.c]: yes

File: mam/api/api.c (hunk 5 above)
Comment (tsvisabo) [reply mam/api/api.c]: woopsi

File: mam/examples/recv.c [hunk 6]
```diff
@@ -37,6 +37,9 @@ int main(int ac, char **av) {
   receive_bundle(av[1], atoi(av[2]), (tryte_t *)av[3], bundle);
 
   mam_psk_t_set_add(&api.psks, &psk);
+  if (ac == 5) {
+    mam_api_add_trusted_channel_pk(&api, av[4]);
```
Comment (thibault-martinez) [inline mam/examples/recv.c]: can you cast to avoid a warning ?

File: mam/mam/message.c [hunk 7]
```diff
@@ -744,30 +750,31 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
...
     if (mam_msg_pubkey_chid1 == pubkey) { /*  SignedId chid1 = 2; */
-      /*TODO: verify chid is trusted */
       ERR_BIND_RETURN(
           mam_msg_unwrap_pubkey_chid1(
               &ctx->spongos, msg, trits_from_rep(MAM_CHANNEL_ID_SIZE, ctx->pk),
               &spongos_mss, &spongos_wots,
               trits_from_rep(MAM_CHANNEL_ID_SIZE, chid)),
           ret);
-      /*TODO: record new channel/endpoint */
+      mam_pk_t_set_add(trusted_channels_pks, chid);
```
Comment (thibault-martinez) [inline mam/mam/message.c]: The added one should be `ctx->pk` ?

File: mam/mam/message.c [hunk 8]
```diff
@@ -744,30 +750,31 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
...
 
-      /*TODO: verify chid is trusted */
       ERR_BIND_RETURN(
           mam_msg_unwrap_pubkey_epid1(
               &ctx->spongos, msg, trits_from_rep(MAM_CHANNEL_ID_SIZE, ctx->pk),
               &spongos_mss, &spongos_wots,
               trits_from_rep(MAM_CHANNEL_ID_SIZE, chid)),
           ret);
-      /*TODO: record new channel/endpoint */
+      mam_pk_t_set_add(trusted_endpoints_pks, chid);
```
Comment (thibault-martinez) [inline mam/mam/message.c]: The added one should be ctx->pk ?

File: mam/mam/message.c [hunk 9]
```diff
@@ -733,6 +735,10 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
                                trits_from_rep(MAM_CHANNEL_ID_SIZE, chid)),
         ret);
     ERR_GUARD_RETURN(0 == ver, RC_MAM_VERSION_NOT_SUPPORTED);
+
+    if (!mam_pk_t_set_contains(trusted_channels_pks, &chid)) {
+      return RC_MAM_PK_IS_NOT_TRUSTED;
+    }
```
Comment (thibault-martinez) [inline mam/mam/message.c:742]: Actually, do we really need to check here ? Do we need chid to be trusted to read a regular message ? Shouldn't we verify that only when engaging in delegation of trust i.e. chid1 or epid1 ?

File: mam/mam/message.c [hunk 10]
```diff
@@ -744,30 +750,31 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
...
+      mam_pk_t_set_add(trusted_endpoints_pks, chid);
     } else if (mam_msg_pubkey_epid ==
                pubkey) { /*  absorb tryte epid[81] = 1; */
       ERR_BIND_RETURN(
           mam_msg_unwrap_pubkey_epid(
               &ctx->spongos, msg, trits_from_rep(MAM_CHANNEL_ID_SIZE, ctx->pk)),
           ret);
+      if (!mam_pk_t_set_contains(trusted_endpoints_pks, &ctx->pk)) {
+        return RC_MAM_PK_IS_NOT_TRUSTED;
+      }
```
Comment (thibault-martinez) [inline mam/mam/message.c:778]: Ditto ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/mam/message.c (hunk 9 above)
Comment (tsvisabo) [reply mam/mam/message.c:742]: you're right

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**